### PR TITLE
feat: add autoware_map_msgs for dynamic map loading

### PR DESCRIPTION
**Pull request by kminoda**
Signed-off-by: kminoda <koji.m.minoda@gmail.com>

## Description
Add autoware_map_msgs to integrate a new interface for dynamic map loading. See https://github.com/orgs/autowarefoundation/discussions/2812  for detail.

## Related links
The next step of this PR would be to create a dynamic map loading functionality in `map_loader`. Here's how it's going to look like: https://github.com/autowarefoundation/autoware.universe/pull/1938

## Notes for reviewers
- Are there any better names for each messages/service?
- Do you think it is redundant to add a `mode` in the response field of `srv/LoadPCDMaps`?

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: autoware_map_msgs/msg/PCDMapWithID.msg
Comment (yukkysaito) [inline autoware_map_msgs/msg/PCDMapWithID.msg:3]: Maybe pointcloud would be better since I think pcd is not common word.
```suggestion
sensor_msgs/PointCloud2 pointcloud
```

Also the file name maybe better 
- PointCloudMapWithID.msg
- IdentifiedPointCloudMap.msg

File: autoware_map_msgs/srv/LoadPCDMaps.srv
Comment (yukkysaito) [inline autoware_map_msgs/srv/LoadPCDMaps.srv:22]: already_loaded_ids means "should be removed map ids"?

File: autoware_map_msgs/msg/PCDMapWithID.msg
Comment (kminoda) [reply autoware_map_msgs/msg/PCDMapWithID.msg]: Thanks! I will take `PointCLoudMapWithID.msg`.
Resolved: https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/05b9d7ffc69e5586a1ad3656045f7179c0051a76

File: autoware_map_msgs/srv/LoadPCDMaps.srv
Comment (yukkysaito) [inline autoware_map_msgs/srv/LoadPCDMaps.srv:3]: I think it is better to separate the service type. How about this?
example:
- PartialPointCloudMap.srv
- DifferentialPointCloudMap.srv

File: autoware_map_msgs/srv/LoadPCDMaps.srv
Comment (kminoda) [reply autoware_map_msgs/srv/LoadPCDMaps.srv:22]: No, it literally means "the map ids that is already loaded", or better to state "the map ids that satisfies the query and is already loaded". Mathematical expression [here](https://github.com/kminoda/autoware_msgs/blob/feat/add_autoware_map_msgs_for_dynamic_map_loading/autoware_map_msgs/README.md#mode-2-differential-map-loading) may help the understandings.

But I think it is also valid to send "ids that should be removed". Do you have any preference on these two?

File: autoware_map_msgs/srv/LoadPCDMaps.srv
Comment (kminoda) [reply autoware_map_msgs/srv/LoadPCDMaps.srv]: I see, thanks. That's much easier to understand.
Resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/ce57ce1baab167dc84d36543e38b3b0aa8b72d31

File: autoware_map_msgs/srv/LoadPCDMaps.srv
Comment (kminoda) [reply autoware_map_msgs/srv/LoadPCDMaps.srv]: I changed the response to "ids that should be removed" at this commit https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/d3706f92918b734cf1736f93222d03ddb26bcb46

File: autoware_map_msgs/srv/PartialPointCloudMap.srv
Comment (yukkysaito) [inline autoware_map_msgs/srv/PartialPointCloudMap.srv:7]: I would like to get from native speakers regarding naming.
cc @mitsudome-r 
example: 
- clear_ids
- removal_ids

File: autoware_map_msgs/srv/DifferentialPointCloudMap.srv
Comment (yukkysaito) [inline autoware_map_msgs/srv/DifferentialPointCloudMap.srv:6]: I would like to get from native speakers regarding naming.
cc @mitsudome-r
example:
- hold_ids
- loaded_ids
- cached_ids

File: autoware_map_msgs/srv/DifferentialPointCloudMap.srv
Comment (yukkysaito) [inline autoware_map_msgs/srv/DifferentialPointCloudMap.srv:10]: [same comment](https://github.com/autowarefoundation/autoware_msgs/pull/39/files#r981952446)
example:
- additional_pointcloud
- additional_pointcloud_with_id
- new_pointcloud
- new_pointcloud_with_id

File: autoware_map_msgs/srv/LoadPCDMaps.srv
Comment (yukkysaito) [reply autoware_map_msgs/srv/LoadPCDMaps.srv]: LGTM :+1:

File: autoware_map_msgs/srv/DifferentialPointCloudMap.srv
Comment (kminoda) [reply autoware_map_msgs/srv/DifferentialPointCloudMap.srv:10]: BTW is "pcd" that unfamiliar? I thought it is pretty common words, given that the extension of the data is ".pcd".

File: autoware_map_msgs/msg/PCDMapWithID.msg
Comment (kenji-miyake) [reply autoware_map_msgs/msg/PCDMapWithID.msg]: Just FYI, other message definitions that have `sensor_msgs/PointCloud2` as a field.

```console
❯ rg ^sensor_msgs/PointCloud2 (fd .msg)
./share/map_msgs/msg/PointCloud2Update.msg
5:sensor_msgs/PointCloud2 points

./share/map_msgs/srv/GetPointMapROI_Response.msg
2:sensor_msgs/PointCloud2 sub_map

./share/map_msgs/srv/GetPointMap_Response.msg
2:sensor_msgs/PointCloud2 map

./share/pcl_msgs/msg/PolygonMesh.msg
4:sensor_msgs/PointCloud2 cloud

./share/pcl_msgs/msg/PolygonMesh.msg
4:sensor_msgs/PointCloud2 cloud

./share/map_msgs/msg/PointCloud2Update.msg
5:sensor_msgs/PointCloud2 points

./share/map_msgs/srv/GetPointMapROI.srv
9:sensor_msgs/PointCloud2 sub_map

./share/map_msgs/srv/GetPointMapROI_Response.msg
2:sensor_msgs/PointCloud2 sub_map

./share/map_msgs/srv/GetPointMap_Response.msg
2:sensor_msgs/PointCloud2 map

./share/map_msgs/srv/GetPointMap.srv
3:sensor_msgs/PointCloud2 map
```

File: autoware_map_msgs/CMakeLists.txt
Comment (kenji-miyake) [inline autoware_map_msgs/CMakeLists.txt:11]: I think `srv` files generally start from a verb.

Comment (kminoda) [reply autoware_map_msgs/CMakeLists.txt:11]: Right, thanks!

1. `LoadDifferentialPointCloudMap.srv` (<- a bit too long)
2. `LoadDifferentialPCDMap.srv` (<- shorter, but depends on whether "pcd" is familiar enough for users)
3. `LoadPCDMapDifferentially.srv` (<- sounds more natural to me)

May I ask your thoughts on this? @kenji-miyake @yukkysaito @mitsudome-r

Comment (kenji-miyake) [reply autoware_map_msgs/CMakeLists.txt:11]: Mm, I don't understand what this service does exactly, but I personally prefer 1. I'll leave it to you.

File: autoware_map_msgs/srv/PartialPointCloudMap.srv
Comment (YamatoAndo) [reply autoware_map_msgs/srv/PartialPointCloudMap.srv:7]: I think it is typo.
It should contain a word that refers to the newly loaded Map.

Comment (kminoda) [reply autoware_map_msgs/srv/PartialPointCloudMap.srv]: Oh, yes, thanks. Fixed https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/ae7ea53556d1f81f8b5c5146afe08118b3538dcf

File: autoware_map_msgs/CMakeLists.txt
Comment (kminoda) [reply autoware_map_msgs/CMakeLists.txt]: Resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/21fa676ce73df4fa56519665e58d00c967aec431

File: autoware_map_msgs/srv/DifferentialPointCloudMap.srv
Comment (kminoda) [reply autoware_map_msgs/srv/DifferentialPointCloudMap.srv]: @mitsudome-r Would you mind giving us some comments on these reviews? (Mainly about the variable names)

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (xmfcx) [inline autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv:6]:
```suggestion
string[] cached_ids
```
Also following from https://github.com/autowarefoundation/autoware_msgs/pull/39#discussion_r981952446

I think `cached_ids` reflects it better.

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (xmfcx) [inline autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv:10]:
```suggestion
PointCloudMapWithID[] new_pointcloud_with_ids
```

Following up from 
https://github.com/autowarefoundation/autoware_msgs/pull/39#discussion_r981953600

Name PCD is not unfamiliar but your message is made of PointCloud2 ros message, which is very different from the [PCD Format](https://pointclouds.org/documentation/tutorials/pcd_file_format.html).

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (xmfcx) [inline autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv:13]: Why are we communicating list of ids to remove?

I think it is better to tell the client what id's to use for the given AreaInfo query instead of the `ids_to_remove`.

Comment (xmfcx) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv:13]: And client can decide what to remove itself, if it wants to.

File: autoware_map_msgs/msg/PointCloudMapWithID.msg
Comment (xmfcx) [inline autoware_map_msgs/msg/PointCloudMapWithID.msg:3]: How about adding
```yaml
geometry_msgs/Point cell_center
geometry_msgs/Vector3 cell_size
```
Also what do you think about adding a center and [geometry_msgs/Vector3](https://github.com/ros2/common_interfaces/blob/rolling/geometry_msgs/msg/Vector3.msg) size information?

This way, the client can reason about this map cell more independently.

And you could name it `CloudMapCellWithId.msg`

File: autoware_map_msgs/msg/PointCloudMapWithID.msg
Comment (xmfcx) [inline autoware_map_msgs/msg/PointCloudMapWithID.msg:2]:
```suggestion
string cell_id
```

**Review by xmfcx — COMMENTED** [review]
I've shared some of my opinions, I can comment more if you agree with `MapCell` idea.

File: autoware_map_msgs/CMakeLists.txt
Comment (xmfcx) [inline autoware_map_msgs/CMakeLists.txt:11]:
```suggestion
  "srv/GetPartialPointCloudMap.srv"
  "srv/GetDifferentialPointCloudMap.srv")
```
Following up from https://github.com/autowarefoundation/autoware_msgs/pull/39#discussion_r986529866

I think `Get` is easier to understand than `Load`. Because the loading is done in the node but the client is calling this service to `Get` the loaded map.

Comment (kminoda) [reply autoware_map_msgs/CMakeLists.txt]: Thanks! Resolved: https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/27a8ea6714aeac6be4f1b394e994aad73fdfa167

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (kminoda) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: Thanks! Resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/27a8ea6714aeac6be4f1b394e994aad73fdfa167

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (kminoda) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/27a8ea6714aeac6be4f1b394e994aad73fdfa167

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (kminoda) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: I and @.yukkysaito had a short discussion on the topic here: https://github.com/autowarefoundation/autoware_msgs/pull/39#discussion_r981918593.

It is true that providing `what id's to use for the given AreaInfo query` would enable the clients adopt more flexible implementation, but we were not sure if this flexibility is currently needed. I believe that it is better to use `ids_to_remove` as long as we don't have any specific use cases in which the client wants `what id's to use` instead of `what id's to remove`.

Do you have any ideas on this? (e.g. any such concrete use case)

Comment (xmfcx) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: - It makes more sense because the client asked for id's in an area and it would return these list of id's. (It's already known by server, easier to implement)
- `what id's to use ` would provide the same functionality as `what id's to remove` but more. Client can choose to remove id's that are not returned.
- If client relies on these responses to remove id's, the server needs to remember what else it has provided to each client, it adds unnecessary complexity.

File: autoware_map_msgs/srv/DifferentialPointCloudMap.srv
Comment (xmfcx) [reply autoware_map_msgs/srv/DifferentialPointCloudMap.srv]: Continued in https://github.com/autowarefoundation/autoware_msgs/pull/39#discussion_r991881747

File: autoware_map_msgs/msg/PointCloudMapWithID.msg
Comment (kminoda) [reply autoware_map_msgs/msg/PointCloudMapWithID.msg]: Resolved: https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/bf074437af0e621f22333d334d1e8b17315c1796

File: autoware_map_msgs/msg/PointCloudMapWithID.msg
Comment (kminoda) [reply autoware_map_msgs/msg/PointCloudMapWithID.msg]: Agreed with the `CellWithID` proposal! https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/bf074437af0e621f22333d334d1e8b17315c1796

Comment (kminoda) [reply autoware_map_msgs/msg/PointCloudMapWithID.msg]: How about this? I think it is better to use this style instead of having center & size, since we can hold a strict boundary information.
```
geometry_msgs/Point min_point
geometry_msgs/Point max_point
```
https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/b5ef05fb21fd4350c132e648707978046de22139

File: autoware_map_msgs/msg/PointCloudMapCellWithID.msg
Comment (xmfcx) [inline autoware_map_msgs/msg/PointCloudMapCellWithID.msg:5]: Are these the boundaries of the point cloud or the cell? (Blue or the red from the image below?)

Untitled

File: autoware_map_msgs/msg/PointCloudMapWithID.msg
Comment (xmfcx) [reply autoware_map_msgs/msg/PointCloudMapWithID.msg]: Sorry didn't see this but could you answer [here](https://github.com/autowarefoundation/autoware_msgs/pull/39#discussion_r991915560)?

File: autoware_map_msgs/msg/PointCloudMapCellWithID.msg
Comment (kminoda) [reply autoware_map_msgs/msg/PointCloudMapCellWithID.msg:5]: I meant the blue one. But now I am not sure whether we should add this boundary information or not.

For the blue type of boundary: Speaking of my dynamic map loading implementation, it is nice to have it in the message, but not necessary (without this boundary info, we should compute min-max again inside `ndt_scan_matcher`, but this is not a big problem for now).

For the red type of boundary: I don't think we need it. It is not useful for `ndt_scan_matcher` at least.

Which type of boundary did you have in mind? @xmfcx

Comment (xmfcx) [reply autoware_map_msgs/msg/PointCloudMapCellWithID.msg:5]: @kminoda I think red makes more sense because it determines the boundaries of the cell.

I don't know if the information about the blue one is used anywhere.

Truth is, since we are making use of a predetermined grid, it should be just enough to share 
- the size of a cell (in meters)
- integer coordinates (origin being 0,0)

This way we wouldn't need an id, we would just share coordinates.

And we would know exactly what cell we are sharing in both the client and the server.

If the `ndt_scan_matcher` uses blue boundary, we can also compute and share it too.

Comment (xmfcx) [reply autoware_map_msgs/msg/PointCloudMapCellWithID.msg:5]: But again, if I am complicating this PR too much, we can just add it the way you intended, and we can think about changing the implementation later on. So please tell if you think it is better to just add it in a way you like.

Comment (kminoda) [reply autoware_map_msgs/msg/PointCloudMapCellWithID.msg:5]: No worries for that. Thanks again for your idea!

> Truth is, since we are making use of a predetermined grid, it should be just enough to share...

I agree that these data is informative enough, and may be necessary if the client wants to have a prior knowledge about the map in order to implement a smarter logic for loading/unloading the map cells. But as far as we've discussed in TIER IV, we could not find any concrete use case in which such boundary information is necessary for the client side.
(FYI: [the proposal A shown in the previous discussion](https://github.com/orgs/autowarefoundation/discussions/2812) also communicates the similar data.)

So my proposal for this comment would be:
- Leave boundary information out of `PointCloudMapCellsWithID`
- Discuss whether to include the boundary information if the needs arise in the future

Comment (xmfcx) [reply autoware_map_msgs/msg/PointCloudMapCellWithID.msg:5]: Ok, let's leave it out then.

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (xmfcx) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: @kminoda any opinions on this?

Comment (xmfcx) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: Also after thinking about this, even `what id's to use` is unnecessary because the client can just iterate through the `PointCloudMapWithID[] loaded_pcds` so I think we can completely remove this item.

Comment (xmfcx) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: @kminoda 

> `loaded_pcds` is merely $X_1 \backslash X_0$, and thus the client would not be able to reconstruct the set $X_0 \backslash X_1$. 

Yeah I was wrong here sorry. But I wanted to say that ( $X_1$ = what IDs to use) and we would provide this back to the client. $X_0 \cap X_1$ = what id's to keep

Untitled

> - Case A (what IDs to remove): The client has $X_0$, $X_1 \backslash X_0$, and $X_0 \backslash X_1$
> - Case B (what IDs to keep): The client has $X_0$, $X_1 \backslash X_0$, and $X_0 \cap X_1$

I will add:
- Case C (queried IDs): The client has $X_0$, $X_1$ with these, complements or intersections can be calculated by the client.

OK so, both A and B would work but you prefer A so that client won't have to do the operation:
- $X_1 \backslash X_0 = X_0 \backslash (X_0 \cap  X_1)$
- (`id's to remove` = `cached ids` - `id's to keep`)

Since they both achieve the same result, we can do it with Case A.

Also with C you can just calculate $X_0 \cap X_1$ too but it's not that important.

Thanks for explaining it.

Comment (kminoda) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: Oh, OK, now I got your point. And thanks for the draw! Then I think we will move on with case A.

FYI, I think Case C would almost will look like Case B in practical. Since we are loading new PCDs only for `new IDs`, the response will look like
- new IDs (all of which attached with a corresponding PCDs)
- IDs to keep
which is the same as Case B.

File: autoware_map_msgs/msg/PointCloudMapCellWithID.msg
Comment (kminoda) [reply autoware_map_msgs/msg/PointCloudMapCellWithID.msg]: @xmfcx 
Sorry I forgot to reflect the conclusion.
Here's the commit: https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/7dd600f7ee8f2c91a6670f56d648b871ab3d1651

**Comment by kminoda** [comment]
I added `frame_id` in the services https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/bfd095d4990d70253a16b75b0dd38978334fd009

File: autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv
Comment (xmfcx) [reply autoware_map_msgs/srv/LoadDifferentialPointCloudMap.srv]: Yes, I will go over the PR one last time and then let's merge it!

File: autoware_map_msgs/README.md
Comment (xmfcx) [inline autoware_map_msgs/README.md:7]:
```suggestion
## PointCloudMapCellWithID.msg
```

File: autoware_map_msgs/README.md
Comment (xmfcx) [inline autoware_map_msgs/README.md:11]:
```suggestion
## GetPartialPointCloudMap.srv
```

File: autoware_map_msgs/README.md
Comment (xmfcx) [inline autoware_map_msgs/README.md:17]:
```suggestion
## GetDifferentialPointCloudMap.srv
```

File: autoware_map_msgs/README.md
Comment (xmfcx) [inline autoware_map_msgs/README.md:5]:
```suggestion
The message represents an area information. This is intended to be used as a query for partial / differential map loading (see `PartialPointCloudMap.srv` section).
```
I guess?

**Review by xmfcx — CHANGES_REQUESTED** [review]

File: autoware_map_msgs/README.md
Comment (kminoda) [reply autoware_map_msgs/README.md]: Resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/43bd07dc583747a68c07afbe8d06fc7f24489d63

File: autoware_map_msgs/README.md
Comment (kminoda) [reply autoware_map_msgs/README.md]: Resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/43bd07dc583747a68c07afbe8d06fc7f24489d63

File: autoware_map_msgs/README.md
Comment (kminoda) [reply autoware_map_msgs/README.md]: Resolved https://github.com/autowarefoundation/autoware_msgs/pull/39/commits/43bd07dc583747a68c07afbe8d06fc7f24489d63

**Comment by kminoda** [comment]
Since I couldn't solve the DCO issue in this PR, I re-opened the same PR in https://github.com/autowarefoundation/autoware_msgs/pull/40

**Comment by kminoda** [comment]
@mitsudome-r @yukkysaito Would you review or approve this PR?

**Review by xmfcx — APPROVED** [review]
Thanks for your efforts, it seems good to me.

**Comment by kminoda** [comment]
@kenji-miyake Sorry, would you make this DCO pass again?

**Comment by kminoda** [comment]
@mitsudome-r @kenji-miyake (mentioning whom seems to have the write access)
Would you merge this PR? I cannot merge by myself since I don't have the write access.

**Comment by kenji-miyake** [comment]
@kminoda I think @mitsudome-r or @xmfcx is the right person to merge this.
(I actually can merge, but I'd like to wait for comments from Mitsudome-san.)

**Comment by xmfcx** [comment]
I've merged it, @mitsudome-r if you have any comments, we can resolve them in another pr.

**Pull request merged**
